cmd/gcp-controller-manager/node_csr_approver.go

```diff
@@ -1007,6 +1019,63 @@ func deleteAllPodsBoundToNode(ctx *controllerContext, nodeName string) error {
 	return utilerrors.NewAggregate(errs)
 }
 
+func markFailedAndDeletePodWithCondition(ctx *controllerContext, pod *v1.Pod, condition *corev1apply.PodConditionApplyConfiguration) error {
+	fieldManager := "GCPControllerManager"
+
+	// Copied directly from pod gc controller (https://github.com/kubernetes/kubernetes/blob/17bf864c1fc20e4badc9a6e0659c5190310462e1/pkg/controller/podgc/gc_controller.go)
+	if utilfeature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) {
+
+		// Extact the pod status as PodGC may or may not own the pod phase, if
+		// it owns the phase then we need to send the field back if the condition
+		// is added.
+		podApply, err := corev1apply.ExtractPodStatus(pod, fieldManager)
+		if err != nil {
+			return nil
+		}
+
+		// Set the status in case PodGC does not own any status fields yet
+		if podApply.Status == nil {
+			podApply.WithStatus(corev1apply.PodStatus())
+		}
+
+		updated := false
+		if condition != nil {
```

Comment:
I think I still need to use `ExtractPodStatus` because it is still desirable to update the pod phase to Failed if the existing status is not Succeeded or Failed (and this requires extracting the pod status). 

Regarding removing condition nil check, I suppose I can remove it, but if someone later calls `markFailedAndDeletePodWithCondition` function and passes a nil condition, it will panic because the `updatePodCondition` will deference the nil condition and panic. As a result, I would prefer to keep the check as is. The other added benefit is the code matches exactly the pod gc controller since we want the same logic here.

Reply:
> I think I still need to use `ExtractPodStatus` because it is still desirable to update the pod phase to Failed if the existing status is not Succeeded or Failed (and this requires extracting the pod status).

Also, you don't want to send the `phase` field unconditionally if it is already in the terminal phase.

> Regarding removing condition nil check, I suppose I can remove it, but if someone later calls `markFailedAndDeletePodWithCondition` function and passes a nil condition, it will panic because the `updatePodCondition` will deference the nil condition and panic. As a result, I would prefer to keep the check as is. The other added benefit is the code matches exactly the pod gc controller since we want the same logic here.

You could pass it as a reference instead of a pointer to prevent the panic, or embed the condition into the function, but I agree that having the function as original makes sense as it makes it clear this is the same logic. I'm fine with both approaches.

Reply:
Updated to inline the condition and removed `updated` flag to simplify the logic.